lib/cusparse/broadcast.jl

```diff
@@ -288,15 +289,18 @@ end
 end
 
 # helpers to index a sparse or dense array
-function _getindex(arg::Union{CuSparseDeviceMatrixCSR,CuSparseDeviceMatrixCSC}, I, ptr)
+@inline function _getindex(arg::Union{CuSparseDeviceMatrixCSR{Tv},CuSparseDeviceMatrixCSC{Tv},CuSparseDeviceVector{Tv}}, I, ptr)::Tv where {Tv}
     if ptr == 0
-        zero(eltype(arg))
+        return zero(Tv)
     else
-        @inbounds arg.nzVal[ptr]
+        return @inbounds arg.nzVal[ptr]::Tv
     end
 end
-_getindex(arg, I, ptr) = Broadcast._broadcast_getindex(arg, I)
 
+@inline function _getindex(arg::CuDeviceArray{Tv}, I, ptr)::Tv where {Tv}
+    return @inbounds arg[I]::Tv
+end
+@inline _getindex(arg, I, ptr) = Broadcast._broadcast_getindex(arg, I)
```

Comment:
Do these `@inline`s help with compilation issues?

Reply:
Also why is the third arg `ptr` here but `is_nnz` in call sites? Can you add comments on some of the behavior here (e.g. the exact preprocessing you're doing in some of the kernels)?

Reply:
Yes can do